kamon-play-2.6.x/src/main/scala/kamon/play/instrumentation/WSInstrumentation.scala

```diff
@@ -22,56 +22,60 @@ import kamon.trace.{Span, SpanCustomizer}
 import kamon.util.CallingThreadExecutionContext
 import org.aspectj.lang.ProceedingJoinPoint
 import org.aspectj.lang.annotation.{Around, Aspect, Pointcut}
-import play.api.libs.ws.StandaloneWSRequest
+import play.api.libs.ws.{StandaloneWSRequest, StandaloneWSResponse, WSRequestExecutor}
 
 import scala.concurrent.Future
 
 @Aspect
 class WSInstrumentation {
 
-  @Pointcut("execution(* play.api.libs.ws.WSRequestExecutor$$anon$2.apply(..)) && args(request)")
-  def onExecuteWSRequest(request: StandaloneWSRequest): Unit = {}
+  @Pointcut("execution(* play.api.libs.ws.ahc.StandaloneAhcWSRequest.filterWSRequestExecutor(..))")
```

Comment:
`StandaloneAhcWSRequest.execute()` was my first bet but while writing these changes but for some reason the instrumentation was running twice: when hitting the pointcut and when calling `pjp.proceed()`. After several attempts hooking on the `WSRequestExecutor` instance seemed the only viable one.

Then I tried the more general version to catch any implementation that might exist but couldn't make it work without double-instrumenting the requests :(.. then tried to see if there are any other implementations to test this and found that one that you mentioned doesn't seem to be maintained anymore:

![image](https://user-images.githubusercontent.com/1302854/38872688-24c38ae4-4254-11e8-8877-9c4d253550e1.png)

It seemed like the only actual, widely used implementation is the one shipped with Play so went with the implementation-specific instrumentation. I'll take a look at it again a bit later to see if I find a general way to solve this, if not I think it will just go this way and we will have to add more instrumentation in the future :disappointed:

Reply:
@ivantopo Thanks for tackling this one! Have you considered putting the instrumentation on `play.api.libs.ws.WSClient.url()`, such that you add a `WSRequestFilter`? I was playing around with this concept just last night in my MDC propagation experiments. I achieve this by wrapping the `WSClient`. Ideally, you could do the same using aspectj, but also expose these component classes in case someone wants to use them outside of aspectj environment.

```
class RequestFilter(implicit ec: ExecutionContext) extends WSRequestFilter {
  def apply(executor: WSRequestExecutor) = WSRequestExecutor { request =>
    val context = MDC.getCopyOfContextMap
    executor(request) andThen {
      case _ => if (context == null) MDC.clear() else MDC.setContextMap(context)
    }
  }
}

object Implicits {
  implicit class TraceableRequest(private val request: WSRequest) extends AnyVal {
    def trace(implicit ec: ExecutionContext): WSRequest = request.withRequestFilter(new RequestFilter)
  }
}

class TraceableWSClient(delegate: WSClient)(implicit ec: ExecutionContext) extends WSClient {
  import Implicits._
  override def underlying[T]: T = delegate.underlying
  override def url(url: String): WSRequest = delegate.url(url).trace
  override def close(): Unit = delegate.close()
}

object TraceableWSClient {
  def apply(delegate: WSClient)(implicit ec: ExecutionContext) = new TraceableWSClient(delegate)
}
```

Reply:
@jsw great suggestion! The changes were minimal and now any `WSClient` implementation should be covered by this instrumentation, thanks a lot! I just pushed the changes and hopefully now this is good to merge. 

@jtjeferreira please take a look as well. Also, regarding this comment:

> but maybe we should instrument the stream method separately...

I created this issue to continue the discussion: https://github.com/kamon-io/kamon-play/issues/33, I think that's a interesting and important topic but lets handle it over there :smiley:

Reply:
Forgot to mention it, [here](https://github.com/kamon-io/kamon-play/pull/32/files#diff-594fa1d7d63a8af242821ccc35823135R47) is the `WSRequestFilter` builder in case someone wants to add it in a non-aspectj environment.